### PR TITLE
fix: fix Clippy warnings

### PR DESCRIPTION
**Pull request by zamazan4ik**
- fix a bunch of Clippy warnings
- fix some usages of assert! (change to assert_ne)

Tested:
- Local unit-tests run

### CODE REVIEW

**Comment by zamazan4ik** [comment]
I didn't fix yet all Clippy warnings since for some of the, that's a little questionable, how to fix them properly.

File: src/compression.rs [hunk 1]
```diff
@@ -134,6 +134,7 @@ mod test {
         }
     }
 
+    #[allow(clippy::vec_init_then_push)]
     fn methods() -> Vec<CompressionMethod> {
```
Comment (Plecra) [inline src/compression.rs:138]: I think clippy wants code like this, which is an alright improvement.
```rust
vec![
    CompressionMethod::Stored
    #[cfg(any(
        feature = "deflate",
        feature = "deflate-miniz",
        feature = "deflate-zlib"
    ))]
    CompressionMethod::Deflated,
    #[cfg(feature = "bzip2")]
    CompressionMethod::Bzip2,
]
```

Comment (zamazan4ik) [reply src/compression.rs:138]: Just was not sure is it nice enough :) Will fix it.

Comment (zamazan4ik) [reply src/compression.rs:138]: As I thought:
```
rror: no rules expected the token `#`
   --> src/compression.rs:140:13
    |
140 |             #[cfg(any(
    |             ^ no rules expected this token in macro call

error: could not compile `zip` due to previous error
warning: build failed, waiting for other jobs to finish...
error: build failed

```

File: src/read.rs [hunk 2]
```diff
@@ -457,14 +457,14 @@ impl<R: Read + io::Seek> ZipArchive<R> {
     }
 
     /// Get a contained file by index
-    pub fn by_index<'a>(&'a mut self, file_number: usize) -> ZipResult<ZipFile<'a>> {
+    pub fn by_index(&mut self, file_number: usize) -> ZipResult<ZipFile> {
```
Comment (Plecra) [inline src/read.rs:460]: Can you add the elided lifetime for `ZipFile<'_>`? The borrow is an important part of the signature and iirc, it affects the docs.

File: src/types.rs [hunk 3]
```diff
@@ -88,6 +88,7 @@ impl DateTime {
     /// * hour: [0, 23]
     /// * minute: [0, 59]
     /// * second: [0, 60]
+    #[allow(clippy::result_unit_err)]
```
Comment (Plecra) [inline src/types.rs:91]: A better error type will be nice when we can fix the api

File: src/types.rs [hunk 4]
```diff
@@ -329,15 +327,15 @@ mod test {
         use super::DateTime;
         let dt = DateTime::default();
         assert_eq!(dt.timepart(), 0);
-        assert_eq!(dt.datepart(), 0b0000000_0001_00001);
+        assert_eq!(dt.datepart(), 0b0000_0000_0010_0001);
```
Comment (Plecra) [inline src/types.rs:330]: The formatting here is useful: `datepart` has three bitfields with different lengths, so using runs of 4 is confusing. (That's the first of January. Not so obvious with the new formatting ;))

**Review by Plecra — CHANGES_REQUESTED** [review]
Marvelous! Clippy's recommendations were mostly great. Even caught some nasty bugs in there, which is nice to see. Only a few style 'regressions' you might want to reconsider.

File: src/types.rs (hunk 3 above)
Comment (zamazan4ik) [reply src/types.rs:91]: Sure. That's why I've commented it out for now :) Later, when we will rework an API, we definitely will fix it (at least I hope)

File: src/compression.rs (hunk 1 above)
Comment (zamazan4ik) [reply src/compression.rs:138]: Nvm, fixed - my mistake

File: src/read.rs (hunk 2 above)
Comment (zamazan4ik) [reply src/read.rs:460]: Done

File: src/compression.rs (hunk 1 above)
Comment (zamazan4ik) [reply src/compression.rs:137]: Fixed

File: src/types.rs (hunk 4 above)
Comment (zamazan4ik) [reply src/types.rs:330]: Fixed

**Review by Plecra — APPROVED** [review]

**Review by Plecra — APPROVED** [review]

**Pull request merged**